storybookjs/storybook
# Feat: Compact folders (smart flatten)

Closes #25014



## What we did



### Feature implementation

In storybook, there is currently a route tree which can go quite deep. With this, it is necessary to click trough all of the folders to access all the files of the project. With compact folders (smart flatten) this problem can be solved.

Compact folders allow folders with just one entry of the same type (relevant for storybook, since there are many types of folders and files) to be collapsed into the same level with single headings, which improves readability and keeps the tree cleaner.

For example: a group folder with just one group folder as entry becomes
a single heading which name is composed by these 2 folders.

Here is a visual example of the feature:

![Alt text](https://i.ibb.co/dmRvGgt/image.png)

### Feature coverage with new unit tests

Since the new feature changes manager API (introduced nested folders types) we find it relevant to add some unit tests to cover this implementation.

## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_



Example:
1. Run storybook in localhost using the command below:
```
$ yarn start
```
2. Verify in the sidebar that folders are now nested

### Documentation



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release



This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_
